nexus/src/app/sagas/region_replacement_drive.rs

```diff
@@ -0,0 +1,1626 @@
+                    "region replacement request {} has new_region_id = None",
+                    params.request.id,
+                )));
+            };
+
+            let new_region: db::model::Region = osagactx
+                .datastore()
+                .get_region(new_region_id)
+                .await
+                .map_err(ActionError::action_failed)?;
+
+            let volume_id = new_region.volume_id().to_string();
+
+            check_from_previous_pantry_step(
+                log,
+                params.request.id,
+                last_request_step.step_time,
+                pantry_address,
+                job_id,
+                &volume_id.to_string(),
+            )
+            .await
+        }
+    }
+}
+
+/// Generate a DriveCheck if the previous step was a Propolis step
+async fn check_from_previous_propolis_step(
+    log: &Logger,
+    request_id: Uuid,
+    step_time: DateTime<Utc>,
+    step_instance_id: Uuid,
+    step_vmm_id: Uuid,
+    instance_and_vmm: InstanceAndActiveVmm,
+) -> Result<DriveCheck, ActionError> {
+    // When this saga recorded a Propolis replacement step, an instance existed
+    // and had a running vmm. Is this true now?
+
+    let Some(current_vmm) = instance_and_vmm.vmm() else {
+        // There is no current VMM, but if the current repair step was
+        // `Propolis` then there was previously one. Some action is required:
+        // namely, attach disk to the pantry and let it perform reconcilation.
+
+        info!(
+            log,
+            "instance from last step no longer has vmm";
+            "region replacement id" => ?request_id,
+            "last replacement drive time" => ?step_time,
+            "last replacement drive step" => "propolis",
+            "instance id" => ?step_instance_id,
+        );
+
+        return Ok(DriveCheck::ActionRequired);
+    };
+
+    // Check if the VMM has changed.
+
+    if current_vmm.id != step_vmm_id {
+        // The VMM has changed! This can be due to a stop and start of the
+        // instance, or a migration. If this is the case, then the new VMM
+        // (propolis server) will be performing reconcilation as part of the
+        // Volume activation. Nexus should be receiving notifications from the
+        // Upstairs there.
+        //
+        // If the new vmm is in the right state, this drive saga can re-send the
+        // target replacement request to poll if the replacement is done yet.
+
+        info!(
+            log,
+            "vmm has changed from last step";
+            "region replacement id" => ?request_id,
+            "last replacement drive time" => ?step_time,
+            "last replacement drive step" => "propolis",
+            "instance id" => ?step_instance_id,
+            "old vmm id" => ?step_vmm_id,
+            "new vmm id" => ?current_vmm.id,
+        );
+
+        Ok(DriveCheck::ActionRequired)
+    } else {
+        // The VMM has not changed: check if the VMM is still active.
```

Comment:
I would consider returning `DriveCheck::Wait` if `instance_and_vmm.instance().runtime().migration_id.is_some()`. `migration_id` gets set at the start of the live migration saga and cleared only when the migration is retired (by one of the participating Propolises/sled agents signaling to Nexus that the migration is over), which covers the point in the migration saga where it checks out the instance's volumes to get the "successor" VCRs to send to the migration target.

If replacements are allowed to proceed while a migration saga is churning, we can end up with races of this sort:

1. Instance starts with VCR V at generation = 20
2. Migration saga checks out the volume, getting V(gen = 21)
3. Before the saga can actually contact the destination sled, a region replacement procedure starts, generating VCR V' with generation 22
4. Replacement sends this VCR to the source Propolis (remember, no migration has started yet!), which will kick off replacement using V'(gen = 22)
5. migration protocol proceeds and tries to activate with V(gen = 21), which is out of date

The check for `VmmState::Migrating` on line 535 doesn't prevent this because the source VMM won't go to the `Migrating` state until the migration protocol actually begins.

It's possible for another migration to start after the first finishes, but I think that's OK here:

1. Propolis P1 starts with VCR V(gen = 20)
2. Migration checks out V(gen = 21)
3. Region replacement start creates V'(gen = 22)
4. Region replacement *drive* is blocked by the ongoing migration
5. Propolis P2 activates with V(gen = 21) and migration completes
6. Drive sends V'(gen = 22) to P2
7. New migration saga begins; this will launch Propolis P3 with V'(gen = 23)

This seems OK--the new Propolis has the latest VCR generation (V') and the latest generation number (23). The point is to use the instance's `migration_id` field to make sure that any task that could have observed the old VCR has concluded before trying to drive replacement with the new VCR.

Reply:
To put the second part another way: it's OK if an instance migration starts after this routine has observed `migration_id.is_none()`, because the new migration is guaranteed to pick up the new VCR version when it goes to check out volumes. (We're relying here on the property that, once a region replacement changes a VCR from V to V', a new replacement won't change it from V' to V'' until the first replacement has finished.)

Reply:
We talked about this in today's storage huddle, and came to the conclusion that this PR shouldn't be blocked but 1) oxidecomputer/crucible#871 needs to be addressed (by allowing it!) and 2) propolis may need a further update.

It's important to note that sending a replacement request does _not_ incur an activation - the first race you pointed out in the first comment here can't happen due to this. Sending the replacement using gen = 22 doesn't activate, and the destination's gen = 21 will still boot out the source's gen = 20.

I agree that the second race you pointed out is ok as long as the replacement request is sent after activation.

68846caf8 will cause the drive saga to bail out when a migration is occurring (although there's a small TOCTOU issue that can occur in the execute node). This may be able to be relaxed in the future if we can prove that the interleaving of these sagas is safe :)